[WebGPU] GPUErrorFilter is missing "internal" value

mwyrzykowski opened the pull request:
#### db45da964d545f2c9b1f0abad275cb539e5bd169
<pre>
[WebGPU] GPUErrorFilter is missing &quot;internal&quot; value
<a href="https://bugs..webkit.org/show_bug.cgi?id=253478">https://bugs..webkit.org/show_bug.cgi?id=253478</a>
&lt;radar://106337546&gt;

Reviewed by Myles C. Maxfield.

Add GPUInternalError and corresponding files.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::popErrorScope):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUDeviceError.idl:
* Source/WebCore/Modules/WebGPU/GPUError.h:
* Source/WebCore/Modules/WebGPU/GPUErrorFilter.h:
(WebCore::convertToBacking):
* Source/WebCore/Modules/WebGPU/GPUErrorFilter.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.cpp:
(WebCore::GPUUncapturedErrorEvent::error const):
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEventInit.idl:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUConvertToBackingContext.cpp:
(PAL::WebGPU::ConvertToBackingContext::convertToBacking):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUErrorFilter.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebKit/Shared/WebGPU/WebGPUConvertFromBackingContext.h:
* Source/WebKit/Shared/WebGPU/WebGPUConvertToBackingContext.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebCore/Modules/WebGPU/GPUInternalError.cpp: Added.
(WebCore::GPUInternalError::message const):
* Source/WebCore/Modules/WebGPU/GPUInternalError.h: Added.
(WebCore::GPUInternalError::create):
(WebCore::GPUInternalError::backing):
(WebCore::GPUInternalError::backing const):
(WebCore::GPUInternalError::GPUInternalError):
* Source/WebCore/Modules/WebGPU/GPUInternalError.idl: Added.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUInternalError.h: Added.
* Source/WebKit/Shared/WebGPU/WebGPUInternalError.cpp: Added.
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUInternalError.h: Added.
Add new files

Canonical link: <a href="https://commits.webkit.org/261545@main">https://commits.webkit.org/261545@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e36ae70a2ee650c9174a4fb8bd2bcb03d831277e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112052 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3848 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120710 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22531 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12332 "Build is in progress. Recent messages:") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3663 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117813 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105114 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/115480 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13600 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/89/builds/482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14278 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/85/builds/12332 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19651 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52481 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8046 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16069 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/Modules/WebGPU/GPUDeviceError.idl:
```diff
@@ -26,7 +26,7 @@
 
 // https://bugs.webkit.org/show_bug.cgi?id=232548 This shouldn't need to be here.
-typedef (GPUOutOfMemoryError or GPUValidationError) GPUError;
+typedef (GPUOutOfMemoryError or GPUValidationError or GPUInternalError) GPUError;
```
The [spec](https://gpuweb.github.io/gpuweb/#gpuerror) says

```
interface GPUError {
    readonly attribute DOMString message;
};
```

Are we sure our IDL is right?

mwyrzykowski replied:
GPUError is a base type according to the specification and we could / probably should implement this as done in the specification.  I think the result is the same however as the base type is never used in an error. To avoid refactoring GPUOutOfMemoryError or GPUValidationError I used the existing pattern which uses a typedef.

mwyrzykowski replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=253744 to track this and will add a FIXME here

litherum commented on Source/WebCore/Modules/WebGPU/GPUInternalError.cpp:
```diff
@@ -0,0 +1,40 @@
+{
+    if (m_backing)
+        return m_backing->message();
+    return m_message;
```
I am confused why there are 2 copies of the message

mwyrzykowski replied:
That's a good question, I followed the pattern as implemented for GPUValidationError from 61ef1f1 - https://bugs.webkit.org/show_bug.cgi?id=232804

Do you happen to know why?

litherum commented on Source/WebCore/Modules/WebGPU/GPUInternalError.cpp:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2023

litherum commented on Source/WebKit/Shared/WebGPU/WebGPUInternalError.cpp:
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2023

mwyrzykowski committed:
[WebGPU] GPUErrorFilter is missing "internal" value
https://bugs..webkit.org/show_bug.cgi?id=253478
<radar://106337546>

Reviewed by Myles C. Maxfield.

Add GPUInternalError and corresponding files.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::popErrorScope):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUDeviceError.idl:
* Source/WebCore/Modules/WebGPU/GPUError.h:
* Source/WebCore/Modules/WebGPU/GPUErrorFilter.h:
(WebCore::convertToBacking):
* Source/WebCore/Modules/WebGPU/GPUErrorFilter.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.cpp:
(WebCore::GPUUncapturedErrorEvent::error const):
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEventInit.idl:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUConvertToBackingContext.cpp:
(PAL::WebGPU::ConvertToBackingContext::convertToBacking):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUErrorFilter.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebKit/Shared/WebGPU/WebGPUConvertFromBackingContext.h:
* Source/WebKit/Shared/WebGPU/WebGPUConvertToBackingContext.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebCore/Modules/WebGPU/GPUInternalError.cpp: Added.
(WebCore::GPUInternalError::message const):
* Source/WebCore/Modules/WebGPU/GPUInternalError.h: Added.
(WebCore::GPUInternalError::create):
(WebCore::GPUInternalError::backing):
(WebCore::GPUInternalError::backing const):
(WebCore::GPUInternalError::GPUInternalError):
* Source/WebCore/Modules/WebGPU/GPUInternalError.idl: Added.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUInternalError.h: Added.
* Source/WebKit/Shared/WebGPU/WebGPUInternalError.cpp: Added.
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUInternalError.h: Added.
Add new files

Canonical link: https://commits.webkit.org/261545@main
